torch/_prims/__init__.py

```diff
@@ -2063,7 +2063,18 @@ def _device_put_meta(
 
 
 def _device_put_aten(a: Tensor, device: Union[str, torch.device]) -> Tensor:
-    return a.to(device)
+    # Make sure that `.to()` doesn't call back into device_put
+    # This can happen because primtorch registered a meta kernel
+    # to `aten._to_copy()`, which calls back into device_put.
+    ctx: Any = contextlib.nullcontext
+    if (
+        device == "meta"
+        or (isinstance(device, torch.device) and device.type == "meta")
+        or a.device == "meta"
+    ):
+        ctx = torch._dispatch.python.no_python_dispatcher
+    with ctx():
+        return a.to(device)
```

Comment:
I guess... I can stamp this to unblock, but this rubs me the wrong way. What's the "Right" semantics here?

Reply:
Hmm, the problem is that our python-only meta implementations are creating a cycle across two meta functions, that we break it in some fashion:

aten._to_copy -> prims.device_put
prims.device_put -> aten._to_copy

_to_copy is implemented as a decomp to prims. So if we want to break the cycle without affecting what the graph looks like, we’ll need to change the meta impl of device_put.

The only way I can think of to make that happen is to get device_put to call our c++ kernel for _to_copy. We’ve already overridden the meta implementation for _to_copy to prefer the python implementation though. So I think our options are:

(1) invoke _to_copy but make sure we don’t hit the Meta key (the c++ kernel is CompositeExplicitAutograd, so technically any of the other backend keys will do what we want).

(2) hit the meta key, but with the python dispatcher turned off (which disables our python impl)

1 is a pain to do because we’d have to remove the Meta key from TLS (similar to how fake tensor adds the meta key to TLS). 2 also felt hacky- I didn’t have much of a preference between them, except we don’t have a python API to do 1

Reply:
It seems, under normal prim design, the correct edge is aten._to_copy to prims.device_put. Then, device_put is obligated to have a manual meta implementation that doesn't make use of anything (and just says what the output device/whatever) should be. Does this make sense?

Reply:
Oh, idk why my brain skipped that and went straight to "find a C++ kernel to call".

You're saying we should just directly create the tensor on the right device, right? That makes sense. Something like `return TensorMeta(same_metadata_as_input_but_new_device)`

Reply:
yeah

Reply:
hmm... this isn't right actually, because I mis-diagnosed.

The problem is [here](https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L1774), in our ref decomp for `_to_copy`.

When we call `torch._prims.device_put(x, device)`, if advertises as  is a CPU tensor, but `device="meta"`, we end up dispatching to the `CPU` dispatch key.

Contrast that to `aten.to(x, device)`, which will hit the `BackendSelect` dispatch key, and then backend select will take care of sending us to the Meta key.

I think the real fix is that `prims.device_put` isn't getting a BackendSelect kernel (because normally codegen takes care of that for aten ops)

Reply:
Ok technically the codegen wouldn't give `prims.device_put` a backend select kernel either, because this op doesn't accept TensorOptions args.

It still seems reasonable to give it a BackendSelect kernel, given the example above though?

Reply:
yeah, backend select sounds legit